editor/src/messages/portfolio/document/document_message_handler.rs

```diff
@@ -640,22 +640,54 @@ impl MessageHandler<DocumentMessage, DocumentMessageContext<'_>> for DocumentMes
 				if any_artboards && parent != LayerNodeIdentifier::ROOT_PARENT {
 					return;
 				}
-
-				// Non-artboards cannot be put at the top level if artboards also exist there
+				// Non-artboards cannot be put at the top level if artboards also exist there.
 				let selected_any_non_artboards = self
 					.network_interface
 					.selected_nodes()
 					.selected_layers(self.metadata())
 					.any(|layer| !self.network_interface.is_artboard(&layer.to_node(), &self.selection_network_path));
-
 				let top_level_artboards = LayerNodeIdentifier::ROOT_PARENT
 					.children(self.metadata())
 					.any(|layer| self.network_interface.is_artboard(&layer.to_node(), &self.selection_network_path));
-
 				if selected_any_non_artboards && parent == LayerNodeIdentifier::ROOT_PARENT && top_level_artboards {
 					return;
 				}
 
+				if as_duplicate {
+					let mut all_new_ids = Vec::new();
+					let selected_layers = self.network_interface.selected_nodes().selected_layers(self.metadata()).collect::<Vec<_>>();
+
+					responses.add(DocumentMessage::DeselectAllLayers);
+					responses.add(DocumentMessage::AddTransaction);
+
+					for selected_layer in selected_layers {
+						let layer_node_id = selected_layer.to_node();
+
+						let mut copy_ids = HashMap::new();
+						copy_ids.insert(layer_node_id, NodeId(0));
+
+						self.network_interface
+							.upstream_flow_back_from_nodes(vec![layer_node_id], &[], network_interface::FlowType::LayerChildrenUpstreamFlow)
+							.enumerate()
+							.for_each(|(index, node_id)| {
+								copy_ids.insert(node_id, NodeId((index + 1) as u64));
+							});
+
+						let nodes: Vec<_> = self.network_interface.copy_nodes(&copy_ids, &[]).collect();
+						let new_ids: HashMap<_, _> = nodes.iter().map(|(id, _)| (*id, NodeId::new())).collect();
+						let layer = LayerNodeIdentifier::new_unchecked(new_ids[&NodeId(0)]);
+						all_new_ids.extend(new_ids.values().cloned());
+
+						responses.add(NodeGraphMessage::AddNodes { nodes, new_ids: new_ids.clone() });
+						responses.add(NodeGraphMessage::MoveLayerToStack { layer, parent, insert_index });
+					}
+
+					responses.add(NodeGraphMessage::RunDocumentGraph);
+					responses.add(NodeGraphMessage::SelectedNodesSet { nodes: all_new_ids });
+
+					return;
+				}
```

Comment:
I have consolidated your new logic into the existing layer move message. Please see if you can further consolidate the logic to share code paths wherever possible. Because currently this just takes your duplicate-ish implementation and pastes it into here (followed by an early `return;` to avoid running the subsequent original non-duplication code paths).